CMakeLists.txt

```diff
@@ -163,6 +163,48 @@ option(BUILD_TESTING "Whether to enable tests" ON)
 
 option(BUILD_W3CTRACECONTEXT_TEST "Whether to build w3c trace context" OFF)
 
+option(MAINTAINER_MODE "Build in maintainer mode (-Wall -Werror)" OFF)
+
+if(MAINTAINER_MODE)
+  add_compile_options(-Wall)
+  add_compile_options(-Werror)
+  add_compile_options(-Wextra)
+
+  if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+
+    # Tested with GCC 9.4 on github.
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 9.4)
+      # Relaxed warnings
+
+      # Enforced warnings
+    endif()
+
+  endif()
+
+  if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+
+    # Tested with Clang 11.0 on github.
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 11.0)
+      # Relaxed warnings
+      add_compile_options(-Wno-error=unused-private-field)
+
+      # Relaxed warning, for GRPC generated code.
+      add_compile_options(-Wno-error=shadow-field)
```

Comment:
Is it better to use include guard to disable these warnings generated by grpc? Just like exporters/otlp/include/opentelemetry/exporters/otlp/protobuf_include_prefix.h and exporters/otlp/include/opentelemetry/exporters/otlp/protobuf_include_suffix.h .So we can also allow users to enable `-Wall -Wextra -Werror` when include headers of otel-cpp.

Reply:
User code should not include grpc generated headers any more, so users can enable `-Wall -Wextra -Werror` now (well, once we cleanup all warnings), independently of what grpc generates.

If we want to enforce -Wshadow-field in the otel-cpp code as well, yes, we will need refinements to disable this for grpc generated code alone. Right now this specific warning is not enforced, for simplicity.

Reply:
For the grpc code in particular, I would rather wait to see if an upgrade to a more recent grpc addresses the warnings, rather than implement workarounds using `protobuf_include_prefix.h` now, for an older grpc/protobuf.

Reply:
> For the grpc code in particular, I would rather wait to see if an upgrade to a more recent grpc addresses the warnings, rather than implement workarounds using `protobuf_include_prefix.h` now, for an older grpc/protobuf.

How about if the application required dependence on a specific grpc? I mean `protobuf_include_prefix.h` should be supported even the same warning is fixed in latest grpc.

Reply:
Now fixed, by using protobuf_include_prefix.h